[PAY-920] K2Q, K2r, and K2RQ factors

mattravenhill-humi opened the pull request:
https://gethumi.atlassian.net/browse/PAY-920

mark-h-humi commented on lib/taxman/taxman2023/calculate.rb:
```diff
@@ -40,6 +40,10 @@ def call
       context[:a] = A.amount(context)
 
+      context[:k2q] = K2Q.amount(context)
+      context[:k2q] = K2R.amount(context)
```
```suggestion
      context[:k2r] = K2R.amount(context)
```

mark-h-humi commented on lib/taxman/taxman2023/k2rq.rb:
```diff
@@ -0,0 +1,50 @@
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2RQ < Factor
```
same re filename/classname

mark-h-humi commented on lib/taxman/taxman2023/k2q.rb:
```diff
@@ -0,0 +1,56 @@
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2Q < Factor
```
same re filename/classname

mark-h-humi commented on lib/taxman/taxman2023/k2r.rb:
```diff
@@ -0,0 +1,38 @@
+      ].min * 0.15
+    end
+
+    def maxium_pension_rate
+      3123.45
+    end
+
+    def maximum_ei
+      1002.45
+    end
+  end
```
constants?

mark-h-humi commented on lib/taxman/taxman2023/k2rq.rb:
```diff
@@ -0,0 +1,50 @@
+      ].min * 0.15
+    end
+
+    def maxium_pension_rate
+      3407.40
+    end
+
+    def maximum_ei
+      781.05
+    end
+
+    def maximum_insurable_earnings
+      449.54
+    end
```
constants

claus-humi commented on lib/taxman/taxman2023/k2r.rb:
```diff
@@ -0,0 +1,38 @@
+      3123.45
+    end
+
+    def maximum_ei
+      1002.45
+    end
```
We can find this value in `Taxman2023::Ei::EI_MAX`. That max is for T4127 so we will need to add two another value in there for QC.

mattravenhill-humi replied:
I've updated the EI constants to include QC_EI_MAX for this purpose

mark-h-humi commented:
also specs please and thank-you @mattravenhill-humi

claus-humi commented on lib/taxman/taxman2023/k2r.rb:
```diff
@@ -0,0 +1,38 @@
+    def pension_amount
+      [
+        p * c * (0.0495 / 0.0595),
+        (maxium_pension_rate * (pm / 12)) - (dq * (0.0495 / 0.0640)) + (dq * (0.0540 / 0.0640))
```
I believe some of these constants are the rates of CPP and QPP.

mattravenhill-humi replied:
Yep, these constants are now referenced by some helper methods that now live in  lib/taxman/taxman2023/k2_generic.rb

claus-humi reviewed (commented):
Generally looks good. We need to use the constants from other classes and we need specs for these classes with some sanity checks until we get scenarios from Greg that includes actual numbers for the whole moving in and out of Quebec thing. This will be a pain to debug in the future since we don't have a reference for these values.

mattravenhill-humi commented on lib/taxman/taxman2023/k2r.rb:
```diff
@@ -0,0 +1,38 @@
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2R < Factor
```
All relevant file names have been updated to be camel-cased

mattravenhill-humi commented on lib/taxman/taxman2023/k2q.rb:
```diff
@@ -0,0 +1,56 @@
+
+    def higher_qpp_rate
+      0.0640
+    end
```
These helper methods now reference constants and live in lib/taxman/taxman2023/k2_generic.rb

mattravenhill-humi commented:
@claus-humi @mark-h-humi 
I've updated the implementation to separate the K2(K2Q/K2R/K2QR) and K2P calculations. 
This is now handled by lib/taxman/taxman2023/k2_generic.rb and lib/taxman/taxman2023/k2_p_generic.rb.

Along with this change I've defined the newly required constants and added some basic specs.

claus-humi reviewed (commented):
I think there's some confusion as to the calculation of K2 as K2 (and K2R/K2Q/K2RQ) are for calculating the federal taxes in T3, and K2P is only used for non QC provinces. It seems strange to me that we are calculating K2 from K2P?

mattravenhill-humi commented on lib/taxman/taxman2023/tax_calculator.rb:
```diff
@@ -43,7 +43,7 @@ def calculate
 
       # Prep factors for T3
-      context[:k2] = K2.amount(context)
+      context[:k2] = K2P.amount(context)
```
Updated to use the proper K2 class

mattravenhill-humi commented on lib/taxman/taxman2023/calculate.rb:
```diff
@@ -40,6 +40,10 @@ def call
+      context[:k2r] = K2R.amount(context)
+      context[:k2rq] = K2RQ.amount(context)
+
       # The non bonus tax calculation should not have the bonus, but we need it for the bonus tax calculation
```
These value calculations have now been moved into the tax_calculator

claus-humi commented on spec/taxman/taxman2023/k2_p_spec.rb:
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
-RSpec.describe Taxman2023::K2P do
```
Spec file should be renamed to reflect class being tested.

mattravenhill-humi replied:
the spec name has been updated

claus-humi reviewed (commented):
Almost there it seems :)
